text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+---
+layout: default
+title: Fabric core block retrieval and signature verification
+nav_order: 3
+---
+
+- Feature Name: Fabric core block retrieval and signature verification
+- Start Date: (fill me in with today's date, 2022-02-01)
+- RFC PR: (leave this empty)
+- Fabric Component: peer, orderer
+- Fabric Issue: (leave this empty)
+
+
+# Table of Contents
+1. [Summary](#summary)
+2. [Motivation](#motivation)
+3. [Existing Approaches](#existing-approaches-for-block-replication-and-block-signature-verification)   
+4. [Proposed Techniques](#proposed-techniques)
+5. [Implementation Plan](#implementation-plan)
+6. [Testing](#testing)
+
+# Summary
+[summary]: #summary
+
+The Hyperledger Fabric core (orderers and peers) are only crash fault tolerant, and cannot handle arbitrary behavior of its ordering service nodes.
+
+
+The current Fabric peers and orderers are implemented to retrieve blocks from a single orderer
+and there is currently no design of how blocks are to be retrieved in the Byzantine setting or how signatures on blocks should be verified.
+
+
+
+Block retrieval from Byzantine orderers needs to tackle denial of service from malicious orderers.
+
+The challenge with block signature validation in the Byzantine setting is avoiding the linear space overhead in the number of certificates that need to be stored as part of the signature information.
+
+While forks of Fabric with a Byzantine Fault Tolerant ordering service [exist](https://github.com/SmartBFT-Go/fabric/), they have not been incorporated into the official Hyperledger Fabric project.
+
+This RFC aims to:
+
+- Give an overview of how denial of service resistant block retrieval and block signature validation is done in SmartBFT.
+- Define how denial of service resistant block retrieval should be conducted in Hyperledger Fabric.
+- Define how Byzantine block signature validation should be constructed and encoded in Hyperledger Fabric.
+
+# Motivation
+[motivation]: #motivation
+
+Currently, Hyperledger Fabric only supports ordering service types that are Crash Fault Tolerant (CFT), not Byzantine Fault Tolerant (BFT).
+
+There are two main attacks that are considered outside of the crash fault tolerant thread model, but must be protected against in the Byzantine fault tolerant setting:
+
+**Block withholding**: A Byzantine orderer might be reluctant to send newly formed blocks to peers that connect to it, either because of malicious intent or to save egress network bandwidth charges from its hosting provider.
+ Fabric architecture relies on peers receiving newly formed blocks as fast as possible, because otherwise client transaction submission times out and smart contract simulation results in stale reads.
```

Comment:
or applying Hanlan's Razor, the nodes are overloaded for example.

The time frame is now defined here, - is this the 'suspicion threshold' mentioned later?

Reply:
No